File: src/runtime/models/minimax_music3/prompt_format.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#include "runtime/models/minimax_music3/prompt_format.h"

#include <algorithm>
#include <cctype>
#include <regex>
#include <sstream>
#include <vector>

namespace trtmc::minimax_music3 {
namespace {

constexpr const char* kImStart = "<|im_start|>";
constexpr const char* kImEnd = "<|im_end|>";
constexpr const char* kCaptionStart = "<|caption_start|>";
constexpr const char* kCaptionEnd = "<|caption_end|>";
constexpr const char* kLyricsStart = "<|lyrics_start|>";
constexpr const char* kLyricsEnd = "<|lyrics_end|>";
constexpr const char* kAudioStart = "<|audio_start|>";

std::vector<std::string> split_lines(const std::string& text) {
    std::vector<std::string> lines;
    std::size_t start = 0;
    while (true) {
        const auto stop = text.find('\n', start);
        if (stop == std::string::npos) {
            lines.push_back(text.substr(start));
            break;
        }
        lines.push_back(text.substr(start, stop - start));
        start = stop + 1;
    }
    return lines;
}

std::string join_lines(const std::vector<std::string>& lines) {
    std::ostringstream joined;
    for (std::size_t index = 0; index < lines.size(); ++index) {
        if (index != 0)
            joined << '\n';
        joined << lines[index];
    }
    return joined.str();
}

std::string replace_all(std::string text, const std::string& from, const std::string& to) {
    if (from.empty())
        return text;
    std::size_t at = 0;
    while ((at = text.find(from, at)) != std::string::npos) {
        text.replace(at, from.size(), to);
        at += to.size();
    }
    return text;
}

} // namespace

std::string clean_caption(const std::string& caption) {
    // <|key value|> becomes "key is value"; a span with no space keeps its
    // inner text.
    static const std::regex special_tag(R"(<\|([^|]*)\|>)");
    std::string text;
    auto begin = std::sregex_iterator(caption.begin(), caption.end(), special_tag);
    auto end = std::sregex_iterator();
    std::size_t last = 0;
    for (auto it = begin; it != end; ++it) {
        const auto match = *it;
        text.append(caption, last, static_cast<std::size_t>(match.position()) - last);
        std::string inner = match[1].str();
        const auto first = inner.find_first_not_of(" \t");
        const auto stop = inner.find_last_not_of(" \t");
        inner = first == std::string::npos ? std::string() : inner.substr(first, stop - first + 1);
        const auto space = inner.find_first_of(" \t");
        text += space == std::string::npos
                    ? inner
                    : inner.substr(0, space) + " is " +
                          inner.substr(inner.find_first_not_of(" \t", space));
        last = static_cast<std::size_t>(match.position() + match.length());
    }
    text.append(caption, last, std::string::npos);

    static const std::regex heading(R"(^\s{0,3}#{1,6}\s+)");
    static const std::regex bullet(R"(^\s*[*+-]\s+)");
    static const std::regex bold(R"(\*\*([^*]+)\*\*)");
    static const std::regex italic(R"((^|[^*])\*([^*\n]+)\*($|[^*]))");
    static const std::regex rule(R"(^\s*[-*_]{3,}\s*$)");

    std::vector<std::string> lines;
    for (auto line : split_lines(text)) {
        line = std::regex_replace(line, heading, "");
        line = std::regex_replace(line, bullet, "");
        while (true) {
            const auto updated = std::regex_replace(line, bold, "$1");
            if (updated == line)
                break;
            line = updated;
        }
        line = std::regex_replace(line, italic, "$1$2$3");
        const auto stop = line.find_last_not_of(" \t\r");
        line = stop == std::string::npos ? std::string() : line.substr(0, stop + 1);
        lines.push_back(std::regex_replace(line, rule, ""));
    }

    text = join_lines(lines);
    text = replace_all(text, "\xe2\x80\xa2 ", "");
    text = replace_all(text, "    ", "");
    static const std::regex blank_runs(R"(\n{2,})");
    return std::regex_replace(text, blank_runs, "\n");
}

std::string normalize_lyrics(const std::string& lyrics) {
    // A line that opens with structure tags keeps only those tags; anything
    // sharing the line goes, which is the contract the model card warns about.
    static const std::regex leading_tags(R"(^[ \t]*((?:\[[^\]]+\][ \t]*)+))");
    std::vector<std::string> kept;
    for (const auto& line : split_lines(lyrics)) {
        std::smatch match;
        if (std::regex_search(line, match, leading_tags)) {
            std::string tags = match[1].str();
            const auto stop = tags.find_last_not_of(" \t");
            kept.push_back(stop == std::string::npos ? std::string() : tags.substr(0, stop + 1));
        } else {
            kept.push_back(line);
        }
    }

    std::string text = join_lines(kept);
    text = replace_all(text, "] ", "]\n");
    text = replace_all(text, " [", "\n[");
    text = replace_all(text, " ^ ", "\n");

    static const std::regex tag(R"(\[([^\]]+)\])");
    std::string lowered;
    auto begin = std::sregex_iterator(text.begin(), text.end(), tag);
    auto end = std::sregex_iterator();
    std::size_t last = 0;
    for (auto it = begin; it != end; ++it) {
        const auto match = *it;
        lowered.append(text, last, static_cast<std::size_t>(match.position()) - last);
        std::string inner = match[1].str();
        std::transform(inner.begin(), inner.end(), inner.begin(),
                       [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
        lowered += "[" + inner + "]";
        last = static_cast<std::size_t>(match.position() + match.length());
    }
    lowered.append(text, last, std::string::npos);
    return "[start]\n" + lowered;
}

std::string assemble_prompt(const std::string& caption, const std::string& lyrics) {
    return std::string(kImStart) + kCaptionStart + clean_caption(caption) + kCaptionEnd +
           kLyricsStart + normalize_lyrics(lyrics) + kLyricsEnd + kImEnd + kAudioStart;
}

} // namespace trtmc::minimax_music3

File: families/minimax_music3/prompt_format.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""The checkpoint's text input contract.

None of this is in the checkpoint: the special-token frame, the two token ids
that classifier-free guidance needs, and the two normalisations applied to the
description and the lyrics all live in the reference implementation
(diffusers v0.40.0, ``modular_pipelines/minimax_music3/encoders.py``, Apache-2.0,
(c) The MiniMax Team and The HuggingFace Team). A runtime that assembles the
prompt differently tokenises to different ids and generates different audio,
so the contract is reproduced here rather than approximated.

The assembled prompt is::

    <|im_start|><|caption_start|>{caption}<|caption_end|>
    <|lyrics_start|>{lyrics}<|lyrics_end|><|im_end|><|audio_start|>

with no separators beyond the tags themselves.
"""

from __future__ import annotations

import re

IM_START, IM_END = "<|im_start|>", "<|im_end|>"
CAPTION_START, CAPTION_END = "<|caption_start|>", "<|caption_end|>"
LYRICS_START, LYRICS_END = "<|lyrics_start|>", "<|lyrics_end|>"
AUDIO_START = "<|audio_start|>"

#: Token ids the reference uses directly rather than resolving through the
#: tokenizer.
AUDIO_END_TOKEN_ID = 151670
AUDIO_CFG_TOKEN_ID = 151654
AUDIO_CODE_OFFSET = 151675

SEMANTIC_VOCAB_SIZE = 16384
MAX_PROMPT_TOKENS = 5000

#: Autoregressive sampling, from the reference.
AR_CFG_SCALE = 1.5
AR_CFG_TOP_K = 50
AR_SAMPLING_TOP_K = 50

_SPECIAL_TAG_RE = re.compile(r"<\|([^|]*)\|>")
_LEADING_TAGS_RE = re.compile(r"^[ \t]*((?:\[[^\]]+\][ \t]*)+)")


class PromptTooLongError(ValueError):
    """Raised when the assembled prompt exceeds the checkpoint's token budget."""


def clean_caption(caption: str) -> str:
    """Normalise a music description to the form the checkpoint expects.

    Rewrites ``<|key value|>`` spans as ``key is value`` and strips the
    markdown the input contract accepts but the model was not trained on.
    """

    def _rewrite_special_tag(match: re.Match) -> str:
        inner = match.group(1).strip()
        parts = inner.split(None, 1)
        return f"{parts[0]} is {parts[1]}" if len(parts) == 2 else inner

    text = _SPECIAL_TAG_RE.sub(_rewrite_special_tag, caption)

    lines: list[str] = []
    for line in text.splitlines():
        line = re.sub(r"^\s{0,3}#{1,6}\s+", "", line)
        line = re.sub(r"^\s*[*+-]\s+", "", line)
        line = re.sub(r"^\s*\*\s+", "", line)
        while "**" in line:
            updated = re.sub(r"\*\*([^*]+)\*\*", r"\1", line)
            if updated == line:
                break
            line = updated
        line = re.sub(r"(?<!\*)\*([^*\n]+)\*(?!\*)", r"\1", line)
        lines.append(line.rstrip())

    text = "\n".join(lines)
    text = re.sub(r"^\s*[-*_]{3,}\s*$", "", text, flags=re.MULTILINE)
    text = text.replace("• ", "").replace("    ", "")
    return re.sub(r"\n{2,}", "\n", text)


def normalize_lyrics(lyrics: str) -> str:
    """Normalise lyrics to the form the checkpoint expects.

    A line that begins with one or more structure tags keeps only those tags;
    any text sharing the line is dropped, which is the contract the model card
    warns about. Tags are lowercased, each is put on its own line, and the
    result is prefixed with ``[start]``.
    """

    kept: list[str] = []
    for line in lyrics.split("\n"):
        match = _LEADING_TAGS_RE.match(line)
        kept.append(match.group(1).strip() if match else line)

    text = "\n".join(kept)
    text = text.replace("] ", "]\n")
    text = text.replace(" [", "\n[")
    text = text.replace(" ^ ", "\n")
    text = re.sub(r"\[([^\]]+)\]", lambda match: f"[{match.group(1).lower()}]", text)
    return f"[start]\n{text}"


def assemble_prompt(caption: str, lyrics: str) -> str:
    """Return the special-token prompt the language model is given."""

    return (
        f"{IM_START}{CAPTION_START}{clean_caption(caption)}{CAPTION_END}"
        f"{LYRICS_START}{normalize_lyrics(lyrics)}{LYRICS_END}{IM_END}{AUDIO_START}"
    )


def check_prompt_length(token_count: int) -> None:
    """Raise when a tokenised prompt exceeds the checkpoint's budget."""

    if token_count > MAX_PROMPT_TOKENS:
        raise PromptTooLongError(
            f"The assembled prompt has {token_count} tokens; "
            f"the maximum is {MAX_PROMPT_TOKENS}"
        )


def unconditional_ids(input_ids: list[int]) -> list[int]:
    """Return the classifier-free counterpart of a tokenised prompt.

    Every token except the first and the two trailing structure tokens is
    replaced by the audio-CFG token, mirroring the reference's
    ``unconditional_ids[:, 1:-2] = AUDIO_CFG_TOKEN_ID``.
    """

    if len(input_ids) < 4:
        raise ValueError(
            f"a prompt needs at least 4 tokens to build its CFG counterpart, "
            f"got {len(input_ids)}"
        )
    head, tail = input_ids[:1], input_ids[-2:]
    body = [AUDIO_CFG_TOKEN_ID] * (len(input_ids) - 3)
    return head + body + tail
